File: Vysakh_HW 3/augmented_lagrangian.m
% Vysakh Venugopal Nonlinear Optimization HW #3 - Problem 3
function [xmin] = augmented_lagrangian(F,c_eq,c_ineq,x0,v0_eq,v0_ineq,r0,beta)
% augmented_lagrangian calculates the minimum of a function constraint by
% equality and inequality constraints


% Error Message if user inout beta is less than one
if beta >= 1
    disp('The penalty growth constant should be less than one');
end
if size(v0_eq,2) ~= 1
    v0_eq = v0_eq';
end
if size(v0_ineq,2) ~= 1
    v0_ineq = v0_ineq';
end
maxloop = 200;
done = false;
loop = 0;
while ~done
    
    % Augmented lagrangian
    F_aug = @(x) F(x) + (1/r0).*(norm(c_eq(x)-((r0/2).*v0_eq))).^2 + (1/r0).*(sum((min(0,(c_ineq(x)-(r0/2).*v0_ineq))).^2));
    
    % Unconstrained optimization
    options = optimoptions(@fminunc,'Display','none','Algorithm','quasi-newton');
    xk = fminunc(F_aug,x0,options);
    
    % v update for equality constraints
    vk_eq = v0_eq - 2*c_eq(xk)/r0;
    
    % v update for inequality constraints
    if c_ineq(xk) < (r0/2) * v0_ineq
        vk_ineq = v0_ineq - ((2/r0)*c_ineq(xk));
    else
        vk_ineq = 0;
    end
    
    % r update
    rk = beta*r0;
    r0 = rk;
    v0_ineq = vk_ineq;
    v0_eq = vk_eq;
    change = x0-xk;
    x0 = xk;
    loop = loop + 1;
    fprintf('Iteration #%i\n',loop);
    
    % Stopping criteria
    if norm(c_eq(xk)) <= 1e-8 && norm(c_ineq(xk)) <= 1e-8
        xmin = xk;
        break;
    end
    if abs(change) <= 1e-4
        xmin = xk;
        break;
    end
    if loop >= maxloop
        break;
    end
end
end

File: VJ_Assignment3/AugmentedLagrangian.py
# Created by Vinay Jakkali on 12/11/2019

"""
--------------------------------------------------------------------------------------
Include code description here:

--------------------------------------------------------------------------------------
"""
import numpy as np
from scipy.optimize import minimize
from autograd import jacobian, hessian

def augmentedlagrangian(F, c_eq, c_ineq, x0, r0=0.25, v0_eq=1, v0_ineq=1, beta=0.1, maxiter=1e3):

    k = 0

    while k < maxiter:

        """ Augmented Lagrangian Function with both Equality and Inequality Constraints """

        C_ineq = c_ineq(x0) - (r0/2)*v0_ineq

        C_ineq = C_ineq[C_ineq < 0]


        F_al_both = lambda x: F(x) + (1/r0) * (np.linalg.norm(c_eq(x) - ((r0/2) * v0_eq)))**2 \
                                  + (1/r0) * sum(C_ineq**2)

        """ Unconstrained Optimization """

        res = minimize(F_al_both, x0, method='Powell', options={'xtol': 1e-8, 'disp': False})
        xk = res.x


        """ Updating v for Equality Constraints """

        vk_eq = v0_eq - ((2/r0) * c_eq(xk))

        """ Updating v for Inequality constraints """

        if c_ineq(xk) < (r0/2) * v0_ineq:

            vk_ineq = v0_ineq - ((2/r0) * c_ineq(xk))

        else:

            vk_ineq = 0


        """ Updating r """

        rk = beta * r0

        """ Updating all values for next iteration """

        r0 = rk
        v0_eq = vk_eq
        v0_ineq = vk_ineq
        delta_x = x0 - xk
        x0 = xk

        """ Iteration number update """
        k += 1

        """ Stopping Criteria """

        if np.linalg.norm(c_eq(x0)) <= 1e-8 and np.linalg.norm(c_ineq(x0)) <= 1e-8:
            return x0

        if abs(np.linalg.norm(delta_x)) <= 1e-5:
            return x0

    return x0

